api/v1alpha4/virtualmachinesnapshot_types.go

```diff
@@ -0,0 +1,142 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package v1alpha4
+
+import (
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// VirtualMachineSnapshotSpec defines the desired state of VirtualMachineSnapshot.
+type VirtualMachineSnapshotSpec struct {
+	// +optional
+	// +kubebuilder:default=false
+	//
+	// Memory represents whether the snapshot includes the VM's
+	// memory. If true, a dump of the internal state of the virtual
+	// machine (a memory dump) is included in the snapshot. Memory
+	// snapshots consume time and resources and thus, take longer to
+	// create.
+	// The virtual machine must support this capability.
+	// When set to false, the power state of the snapshot is set to
+	// false.
+	// For a VM in suspended state, memory is always included
+	// in the snashot.
+	Memory *bool `json:"memory,omitempty"`
+
+	// +optional
+	// +kubebuilder:default=false
+
+	// Quiesce represents the spec used for granular control over
+	// quiesce details. If quiesceSpec is set and the virtual machine
+	// is powered on when the snapshot is taken, VMware Tools is used
+	// to quiesce the file system in the virtual machine. This assures
+	// that a disk snapshot represents a consistent state of the guest
+	// file systems. If the virtual machine is powered off or VMware
+	// Tools are not available, the quiesce spec is ignored.
+	QuiesceSpec *QuiesceSpec `json:"quiesce,omitempty"`
+
+	// +optional
+	//
+	// Description represents a description of the snapshot.
+	Description string `json:"description,omitempty"`
+
+	// VMRef represents the name of the virtual machine for which the
+	// snapshot is requested.
+	VMRef corev1.TypedLocalObjectReference `json:"vmRef:omitempty"`
+}
+
+// QuiesceSpec represents specifications that will be used to quiesce
+// the guest when taking a snapshot.
+type QuiesceSpec struct {
+	// +optional
+
+	// Timeout represents the maximum time in minutes for snapshot
+	// operation to be performed on the virtual machine. The timeout
+	// can not be less than 5 minutes or more than 240 minutes.
+	Timeout *time.Duration `json:"timeout,omitempty"`
+}
+
+// VMSnapshotPhase represents the phase of a VM snapshot.
+type VMSnapshotPhase string
+
+const (
+	// VM snapshot operation is successful.
+	VMSnapshotSucceeded VMSnapshotPhase = "Succeeded"
+	// VM snapshot operation is in progress.
+	VMSnapshotInProgress VMSnapshotPhase = "In Progress"
+	// VM snapshot operation has failed.
+	VMSnapshotFailed VMSnapshotPhase = "Failed"
+)
+
+const (
+	// VirtualMachineSnapshotReadyCondition represents the condition
+	// that the virtual machine snapshot is ready.
+	VirtualMachineSnapshotReadyCondition = "VirtualMachineSnapshotReady"
+)
+
+// VirtualMachineSnapshotStatus defines the observed state of VirtualMachineSnapshot.
+type VirtualMachineSnapshotStatus struct {
+	// +optional
+
+	// Phase represents the phase of a VM snapshot operation.
+	Phase *VMSnapshotPhase `json:"phase:omitempty"`
```

Comment:
I think we can do without a pointer here?. I referred to definitions for Phase in Persistent Volumes.